### PR TITLE
Fix allocation in TrilinosWrappers::SparseMatrix::AccessorBase

### PR DESCRIPTION
**Pull request by peterrum**
FYI @marcfehling This should cut down the setup costs by a factor 20 or so.

### CODE REVIEW

File: source/lac/trilinos_sparse_matrix.cc [hunk 1]
```diff
@@ -180,6 +180,7 @@ namespace TrilinosWrappers
           colnum_cache->data()));
       value_cache->resize(ncols);
       colnum_cache->resize(ncols);
```
Comment (masterleinad) [inline source/lac/trilinos_sparse_matrix.cc:182]: If we know that `ncols==colnums` then we don't need these two lines anymore.

Comment (peterrum) [reply source/lac/trilinos_sparse_matrix.cc]: You are right :+1:

**Comment by marcfehling** [comment]
> This should cut down the setup costs by a factor 20 or so.

More like a factor of 200 :) That patch makes a huge difference. Thank you!

Current timings with this patch:
```
| Section                      | no. calls |   min time  rank |   avg time |   max time  rank |
+------------------------------------------+------------------+------------+------------------+
| full_cycle                   |         1 |     26.57s    34 |     26.58s |     26.58s    92 |
|  solve                       |         1 |     23.31s    13 |     23.32s |     23.33s    95 |
|   initialize_extdiag()       |        16 |     1.069s    55 |     1.264s |     1.724s    95 |
|    extract_remote_rows()     |        16 |    0.9754s    79 |    0.9786s |    0.9808s    17 |
|     sparse_matrix.begin(row) |  overflow |    0.1544s    95 |    0.2341s |    0.3059s    53 | <---
```

Previous timings without this patch:
```
| Section                      | no. calls |   min time  rank |   avg time |   max time  rank |
+------------------------------------------+------------------+------------+------------------+
| full_cycle                   |         1 |     221.1s    76 |     221.1s |     221.1s    92 |
|  solve                       |         1 |     217.8s    29 |     217.8s |     217.8s    95 |
|   initialize_extdiag()       |        16 |     195.3s    55 |     195.5s |     196.0s    95 |
|    extract_remote_rows()     |        16 |     195.2s    90 |     195.2s |     195.2s    17 |
|     sparse_matrix.begin(row) |  overflow |    0.1532s    95 |     44.47s |     114.6s    42 | <---
```

**Review by marcfehling — APPROVED** [review]

**Comment by bangerth** [comment]
Very nice!

**Comment by bangerth** [comment]
@marcfehling pointed this patch out to me again, and that made me want to look at whether the PETSc and the new Trilinos Tpetra wrappers have the same problem. It turns out that the answer is no, but I thought that's worth documenting here.

**Pull request merged**